src/mumble/AudioOutput.cpp

```diff
@@ -168,35 +183,69 @@ void AudioOutput::addFrameToBuffer(ClientUser *sender, const Mumble::Protocol::A
 
 		qrwlOutputs.lockForWrite();
 
-		aop = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
-		qmOutputs.replace(sender, aop);
+		speech = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
+		qmOutputs.replace(sender, speech);
 	}
 
-	aop->addFrameToBuffer(audioData);
+	speech->addFrameToBuffer(audioData);
 
 	qrwlOutputs.unlock();
 }
 
-void AudioOutput::removeBuffer(const ClientUser *user) {
-	removeBuffer(qmOutputs.value(user));
-}
-
-void AudioOutput::removeBuffer(AudioOutputUser *aop) {
+void AudioOutput::handleInvalidatedBuffer(AudioOutputBuffer *buffer) {
 	QWriteLocker locker(&qrwlOutputs);
-	QMultiHash< const ClientUser *, AudioOutputUser * >::iterator i;
+	QMultiHash< const ClientUser *, AudioOutputBuffer * >::iterator i;
 	for (i = qmOutputs.begin(); i != qmOutputs.end(); ++i) {
-		if (i.value() == aop) {
+		if (i.value() == buffer) {
 			qmOutputs.erase(i);
-			delete aop;
+			delete buffer;
+			break;
+		}
+	}
+}
+
+void AudioOutput::handlePositionedBuffer(AudioOutputBuffer *buffer, float x, float y, float z) {
+	QReadLocker locker(&qrwlOutputs);
+	QMultiHash< const ClientUser *, AudioOutputBuffer * >::iterator i;
+	for (auto iter = qmOutputs.begin(); iter != qmOutputs.end(); ++iter) {
+		if (*iter == buffer) {
+			buffer->fPos[0] = x;
+			buffer->fPos[1] = y;
+			buffer->fPos[2] = z;
 			break;
 		}
 	}
 }
 
-AudioOutputSample *AudioOutput::playSample(const QString &filename, float volume, bool loop) {
+void AudioOutput::setBufferPosition(const AudioOutputToken &token, float x, float y, float z) {
+	if (token.m_buffer == nullptr) {
+		return;
+	}
+
+	emit bufferPositionChanged(token.m_buffer, x, y, z);
+}
+
+void AudioOutput::removeBuffer(AudioOutputBuffer *buffer) {
+	if (buffer == nullptr) {
```

Comment:
```suggestion
	if (!buffer) {
```
this is the idiomatic way (in C++) to perform this check. The upside is that this will keep working even when switching the raw pointer for e.g. a smart pointer